src/artemis/devices/oav/oav_centring.py

```diff
@@ -0,0 +1,692 @@
+        x_positions = np.array([], dtype=np.int32)
+        y_positions = np.array([], dtype=np.int32)
+        widths = np.array([], dtype=np.int32)
+        omega_angles = np.array([], dtype=np.int32)
+        mid_lines = np.array([], dtype=np.int32)
+        tip_x_positions = np.array([], dtype=np.int32)
+        tip_y_positions = np.array([], dtype=np.int32)
+
+        for i in range(rotations):
+            current_omega, top, bottom, tip_x, tip_y, omega_limit = tuple(
+                self.extract_data_from_pvs()
+            )
+            (x, y, width, mid_line) = self.find_midpoint(top, bottom)
+
+            # Build arrays of edges and width, and store corresponding gonomega
+            x_positions = np.append(x_positions, x)
+            y_positions = np.append(y_positions, y)
+            widths = np.append(widths, width)
+            omega_angles = np.append(omega_angles, current_omega)
+            mid_lines = np.append(mid_lines, mid_line)
+            tip_x_positions = np.append(tip_x_positions, tip_x)
+            tip_y_positions = np.append(tip_y_positions, tip_y)
+
+            # rotate the pin to take next measurement, unless it's the last measurement
+            if i < rotations - 1:
+                yield from bps.mv(self.oav_goniometer.omega, current_omega + increment)
+
+        yield (
+            x_positions,
+            y_positions,
+            widths,
+            omega_angles,
+            mid_lines,
+            tip_x_positions,
+            tip_y_positions,
+        )
+
+    def find_widest_point_and_orthogonal_point(
+        x_positions, y_positions, widths, omega_angles
+    ):
+        """
+        Find the widest point from the sampled positions, and the angles orthogonal to this.
+
+        Args: Lists of values taken, the ith value of the list is the ith point sampled:
+                x_positions: the x positions of centres
+                y_positions: the y positions of centres
+                widths: the widths between the top and bottom waveforms at the centre point
+                omega_angles: the angle of the goniometer at which the measurement was taken
+                mid_lines: the waveform going between the top and bottom waveforms
+                tip_x_positions: the measured x tip at a given rotation
+                tip_y_positions: the measured y tip at a given rotation
+        Returns: The index  of the sample which is wildest.
+        """
+
+        # filter out 0 elements
+        zero_x_positions = np.where(x_positions == 0)[0]
+        zero_y_positions = np.where(y_positions == 0)[0]
+        indices_to_remove = np.union1d(zero_x_positions, zero_y_positions)
+        x_positions_filtered = np.delete(x_positions, indices_to_remove)
+        y_positions_filtered = np.delete(y_positions, indices_to_remove)
+        widths_filtered = np.delete(widths, indices_to_remove)
+        omega_angles_filtered = np.delete(omega_angles, indices_to_remove)
```

Comment:
To find outliers, GDA averages the x positions and if it is outside the mean x_positions by 100 then that reading is disregarded as a possible maximum width. I would suppose GDA is doing this as places where it's 0 will almost certainly be outliars, but would skew the median to a great degree.

We need to remove those indices  from the other lists so that the ith index of widths_filtered corresponds to the ith of x_positions_filtered - I think I should actually make a class for the pin data collected from a single rotation. Then use functions to do this kind of thing..